File: ImplementTrie(PrefixTree).cpp
class TrieNode
{
public:
    TrieNode *child[26];
    bool isWord;
    
    TrieNode()
    {
        for(int i = 0;i < 26;i++)child[i]= nullptr;
        isWord = false;
    }
};

class Trie
{
    TrieNode *root;
public:
    Trie()
    {
        root = new TrieNode();
    }

    void insert(string s)
    {
        TrieNode *p = root;
        for(int i = 0; i < s.size(); ++ i)
        {
            if(p -> child[s[i] - 'a'] == nullptr)
                p -> child[s[i] - 'a'] = new TrieNode();
            p = p -> child[s[i] - 'a'];
        }
        p -> isWord = true;
    }

    bool search(string key)
    {
        TrieNode *p = find(key);
        return p != nullptr && p -> isWord;
    }

    bool startsWith(string prefix)
    {
        return find(prefix) != nullptr;
    }

private:
    TrieNode* find(string key)
    {
        TrieNode *p = root;
        for(int i = 0; i < key.size() && p != nullptr; ++ i)
            p = p -> child[key[i] - 'a'];
        return p;
    }
};

File: ImplementTrie(PrefixTree).java
class TrieNode{
    public TrieNode[] child =  new TrieNode[26];
    public boolean flag;

    public TrieNode(){
        flag = false;
        for(int i = 0;i < 26;i++)child[i]=null;
    }
}

class Trie {

    TrieNode root;
    public Trie() {
        this.root = new TrieNode();
    }

    public void insert(String word) {
        TrieNode r = root;
        for(int i = 0;i < word.length();i++){
            if(r.child[word.charAt(i)-'a'] == null)
                r.child[word.charAt(i) - 'a'] = new TrieNode();
            r = r.child[word.charAt(i) - 'a'];
        }
        r.flag = true;
    }
    
    public boolean search(String word) {
        TrieNode r=root;
        for(int i = 0;i < word.length();i++){
            if(r == null)return false;
            r = r.child[word.charAt(i) - 'a'];
        }
        return r != null && r.flag == true;
    }
    
    public boolean startsWith(String prefix) {
        TrieNode r = root;
        for(int i = 0;i < prefix.length();i++){
            if(r == null)return false;
            r = r.child[prefix.charAt(i)-'a'];
        }
        return r!=null;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */
